File: dynabuffers-python/dynabuffers/Dynabuffers.py
from antlr4 import *

from dynabuffers.DynabuffersEngine import DynabuffersEngine
from dynabuffers.DynabuffersVisitor import DynabuffersVisitor
from dynabuffers.antlr.DynabuffersLexer import DynabuffersLexer
from dynabuffers.antlr.DynabuffersParser import DynabuffersParser


class Dynabuffers:
    @staticmethod
    def parse(schema) -> DynabuffersEngine:
        lexer = DynabuffersLexer(Dynabuffers.stream(schema))
        stream = CommonTokenStream(lexer)
        parser = DynabuffersParser(stream)

        visitor = DynabuffersVisitor()
        tree = visitor.visit(parser.compilation())

        return DynabuffersEngine(tree)

    @staticmethod
    def stream(obj):
        if isinstance(obj, InputStream):
            return obj
        if isinstance(obj, str):
            return InputStream(str(obj))
        if isinstance(obj, bytes):
            return InputStream(obj.decode("utf-8"))
        if isinstance(obj, bytearray):
            return InputStream(obj.decode("utf-8"))
        if isinstance(obj, DynabuffersResource):
            return obj.get_content_stream()
        raise ValueError(str(obj) + " cannot be converted to input stream")


class DynabuffersResource:

    def __init__(self):
        self.builder = ""

    def append(self, schema):
        if isinstance(schema, str):
            self.builder += str(schema)
        if isinstance(schema, bytes):
            self.builder += schema.decode("utf-8")
        if isinstance(schema, bytearray):
            self.builder += schema.decode("utf-8")
        raise ValueError(str(schema) + " cannot be appended to resource")

    def get_content_stream(self):
        return InputStream(self.builder)

File: dynabuffers-java/src/main/kotlin/dynabuffers/Dynabuffers.kt
package dynabuffers

import dynabuffers.antlr.DynabuffersLexer
import dynabuffers.antlr.DynabuffersParser
import dynabuffers.api.IType
import dynabuffers.exception.DynabuffersException
import dynabuffers.exception.DynabuffersExceptionListener
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CodePointCharStream
import org.antlr.v4.runtime.CommonTokenStream
import java.io.InputStream
import java.io.Reader
import java.nio.channels.ReadableByteChannel
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets.UTF_8
import java.util.*
import java.util.Optional.empty

/**
 * Factory class for creating a DynabuffersEngine instance.
 */
class Dynabuffers {

    companion object {

        @JvmStatic
        @JvmOverloads
        fun parse(string: String, namespace:Optional<String> = empty<String>(), charset: Charset = UTF_8): DynabuffersEngine {
            return DynabuffersEngine(parse(CharStreams.fromString(string), charset))
        }

        @JvmStatic
        @JvmOverloads
        fun parse(stream: InputStream, charset: Charset = UTF_8): DynabuffersEngine {
            return DynabuffersEngine(parse(CharStreams.fromStream(stream), charset))
        }

        @JvmStatic
        @JvmOverloads
        fun parse(reader: Reader, charset: Charset = UTF_8): DynabuffersEngine {
            return DynabuffersEngine(parse(CharStreams.fromReader(reader), charset))
        }

        @JvmStatic
        @JvmOverloads
        fun parse(channel: ReadableByteChannel, charset: Charset = UTF_8): DynabuffersEngine {
            return DynabuffersEngine(parse(CharStreams.fromChannel(channel), charset))
        }

        @JvmStatic
        @JvmOverloads
        fun parse(resource: DynabuffersResource, charset: Charset = UTF_8): DynabuffersEngine {
            return DynabuffersEngine(parse(resource.getCharStream(), charset))
        }

        private fun parse(stream: CharStream, charset: Charset): List<IType> {
            val lexer = DynabuffersLexer(stream)
            val commonTokenStream = CommonTokenStream(lexer)
            val parser = DynabuffersParser(commonTokenStream)

            val errorListener = DynabuffersExceptionListener()
            parser.addErrorListener(errorListener)

            val visitor = DynabuffersVisitor(charset)
            val astList = visitor.visit(parser.compilation())
                    ?: throw DynabuffersException("invalid dynabuffers scheme")

            if (errorListener.get() != null) {
                throw DynabuffersException(errorListener.get())
            }
            return astList
        }
    }

    /**
     * This class can be used to group some resources for the dynabuffers parsing.
     */
    class DynabuffersResource {

        private val builder = StringBuilder()

        fun append(string: String) {
            builder.append(string)
        }

        @JvmOverloads
        fun append(stream: InputStream, charset: Charset = UTF_8) {
            builder.append(stream.reader(charset).readText())
        }

        fun append(reader: Reader) {
            builder.append(reader.readText())
        }

        fun getCharStream(): CodePointCharStream = CharStreams.fromString(builder.toString())

    }

}
